nexus/src/app/sagas/instance_create.rs

```diff
@@ -705,39 +705,90 @@ async fn sic_allocate_instance_external_ip_undo(
 }
 
 /// Create disks during instance creation, and return a list of disk names
-// TODO implement
 async fn sic_create_disks_for_instance(
     sagactx: NexusActionContext,
 ) -> Result<Option<String>, ActionError> {
+    let osagactx = sagactx.user_data();
     let disk_params = sagactx.saga_params::<DiskParams>()?;
     let saga_params = disk_params.saga_params;
     let disk_index = disk_params.which;
     let saga_disks = &saga_params.create_params.disks;
+    let opctx =
+        OpContext::for_saga_action(&sagactx, &saga_params.serialized_authn);
 
     if disk_index >= saga_disks.len() {
         return Ok(None);
     }
 
     let disk = &saga_disks[disk_index];
 
+    // TODO-correctness TODO-security It's not correct to re-resolve the
+    // organization and project names now.  See oxidecomputer/omicron#1536.
+    let organization_name: db::model::Name =
+        saga_params.organization_name.clone().into();
+    let project_name: db::model::Name = saga_params.project_name.clone().into();
+
     match disk {
-        params::InstanceDiskAttachment::Create(_create_params) => {
-            return Err(ActionError::action_failed(
-                "Creating disk during instance create unsupported!".to_string(),
-            ));
+        params::InstanceDiskAttachment::Create(create_params) => {
+            osagactx
+                .nexus()
+                .project_create_disk(
+                    &opctx,
+                    &organization_name,
+                    &project_name,
+                    &create_params,
+                )
+                .await
+                .map_err(ActionError::action_failed)?;
```

Comment:
I *think* this may have some idempotency problems as-is -- I also wanna tag in @andrewjstone who actually wrote the subsaga implementation, so please, correct me if I'm wrong.

**TL;DR: I think that you are creating two totally distinct sagas here, rather than actually registering the disk creation as a sub-saga**

AFAICT, this is what you are currently doing in this PR:
- Constructing a [subsaga for each createable disk](https://github.com/oxidecomputer/omicron/blob/4be42b45142f58799c6a5444370bbdd75c0340de/nexus/src/app/sagas/instance_create.rs#L293-L316), which performs the "create" and "attach" steps as separate saga steps.
- Within the "create" step, we call `osagactx.nexus().project_create_disk(...)`. This validates the creation parameters, but mostly [runs the `DiskCreate` saga](https://github.com/oxidecomputer/omicron/blob/4be42b45142f58799c6a5444370bbdd75c0340de/nexus/src/app/disk.rs#L213-L225). I **THINK** this is a slightly unconventional approach to subsagas - rather than actually registering the disk creation steps with the "instance creation" saga, this "create disks" saga node is a *single* node. Yes, it *happens* to call code that launches another saga, but the "instance create" saga has no idea that's happening, so it'll retry the entire node (launching a new saga to create a disk) on failure.

Let's consider the following scenario:
1. We start an instance creation saga, with a request to make a disk named "foobar" as part of the saga
2. We get to this saga node, and launch the disk creation sub-saga to actually create the "foobar" disk.
3. Let's suppose we get most of the way through disk creation, but we fail at some point because Nexus crashes.

On reboot, the following could happen:
- The "disk create" saga could get replayed first. In that case, we would successfully create the disk! However, this "instance create" saga is totally independent, so this node (meaning `sic_create_disks_for_instance`) will be called again. On the second invocation of `project_create_disk`, we'll see that a disk already exists named "foobar", so we would fail. Even worse, we'd unwind the instance creation saga without trying to delete the disk which was created -- it was created in a separate saga!
- The "instance create" saga could get replayed first. In this case, we'd launch an *entirely new* saga by calling `project_create_disk` again, even though we already started doing one before Nexus crashed (so it'll still be attempted in the saga replay). This *might* work, or it *might* fail, depending on how far the other saga got before bailing. Either way, there's a decent chance to leak a disk here, since it'll *also* try to be replayed independently of this saga.

Reply:
I mentioned this in chat, but for posterity:

I think it may be easier to think about this problem if we view it as "construct a single DAG for the instance create saga, that consists of the disk creation" rather than "create a DAG for the instance create saga that *might* spawn totally distinct DAGs for disk creation".

To do this, I'd recommend re-using the [disk_create](https://github.com/oxidecomputer/omicron/blob/4be42b45142f58799c6a5444370bbdd75c0340de/nexus/src/app/sagas/disk_create.rs#L76-L79) DAG construction method, and using it to create a [`NodeKind::Subsaga`](https://github.com/oxidecomputer/steno/blob/53ee324036d901649718d32acfb4d238fcaf8f20/src/dag.rs#L222) which gets appended somewhere in the instance create DAG

Reply:
@smklein is totally correct here. The scenario laid out, and the current implementation here is exactly how subsagas originally worked, which is what made them unsound. A subsaga was just a single node that spawned a new saga. Now, as Sean points out, what steno does is take a `Node` with `NodeKind::Subsaga` and when that node gets appended the Dag that is the subsaga is traversed and the nodes are inserted in the parent saga so that they are run in the parent saga. In essence, every subsaga DAG is merged into the parent DAG. 

Practically speaking, what this means is that you can't use the [subsaga_append](https://github.com/oxidecomputer/omicron/blob/4be42b45142f58799c6a5444370bbdd75c0340de/nexus/src/app/sagas/instance_create.rs#L309-L315) call with the [disk_create (`make_saga_dag`)] DAG construction method @smklein suggested because that helper function takes a `DagBuilder` for the child saga and not a `DAG`.

So what you need to do is essentially combine the two methods above, such that you append the subsaga Params to the parent saga as a constant node, and then you append the `Dag` returned from [`make_saga_dag`][https://github.com/oxidecomputer/omicron/blob/4be42b45142f58799c6a5444370bbdd75c0340de/nexus/src/app/sagas/disk_create.rs#L76] as a subsaga node. There's an [example](https://github.com/oxidecomputer/steno/blob/main/src/example_provision.rs#L171-L234) in steno that may help you. But it's essentially what happens in `subsaga_append`.

Reply:
I think there's a similar example in the remove read-only parent saga:
https://github.com/oxidecomputer/omicron/blob/4be42b45142f58799c6a5444370bbdd75c0340de/nexus/src/app/sagas/volume_remove_rop.rs#L78-L89
https://github.com/oxidecomputer/omicron/blob/4be42b45142f58799c6a5444370bbdd75c0340de/nexus/src/app/sagas/volume_remove_rop.rs#L113-L129

Reply:
I paired with @andrewjstone earlier this evening so I think I have a decent grasp of how to resolve this (and clean up some of the linger patterns)